src/main/java/io/aiven/kafka/auth/VerdictCache.java

```diff
@@ -16,44 +16,77 @@
 
 package io.aiven.kafka.auth;
 
+import javax.annotation.Nonnull;
+
+import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 import org.apache.kafka.common.security.auth.KafkaPrincipal;
 
+import io.aiven.kafka.auth.json.AclPermissionType;
 import io.aiven.kafka.auth.json.AivenAcl;
 
 public class VerdictCache {
-    private final List<AivenAcl> aclEntries;
+    private final List<AivenAcl> allowAclEntries;
+    private final List<AivenAcl> denyAclEntries;
     private final Map<String, Boolean> cache = new ConcurrentHashMap<>();
 
-    private VerdictCache(final List<AivenAcl> aclEntries) {
-        this.aclEntries = aclEntries;
+    private VerdictCache(@Nonnull final List<AivenAcl> denyAclEntries, @Nonnull final List<AivenAcl> allowAclEntries) {
+        this.denyAclEntries = denyAclEntries;
+        this.allowAclEntries = allowAclEntries;
     }
 
     public boolean get(final KafkaPrincipal principal,
                        final String operation,
                        final String resource) {
-        if (aclEntries != null) {
-            final String cacheKey = resource
-                + "|" + operation
-                + "|" + principal.getName()
-                + "|" + principal.getPrincipalType();
-            final Predicate<AivenAcl> matcher = aclEntry ->
-                aclEntry.check(principal.getPrincipalType(), principal.getName(), operation, resource);
-            return cache.computeIfAbsent(cacheKey, key -> aclEntries.stream().anyMatch(matcher));
+        final String principalType = principal.getPrincipalType();
+        final String cacheKey = resource
+            + "|" + operation
+            + "|" + principal.getName()
+            + "|" + principalType;
+
+        final Boolean result = cache.get(cacheKey);
+        if (result != null) {
+            return result;
+        }
+
+        final Predicate<AivenAcl> matcher = allowAcl ->
+            allowAcl.match(principalType, principal.getName(), operation, resource);
+
+        if (!denyAclEntries.isEmpty()) {
+            // check if it's denied
+            final boolean denied = denyAclEntries.stream().anyMatch(matcher);
+            if (denied) {
+                cache.put(cacheKey, false);
+                return false;
+            }
+        }
+
+        if (!allowAclEntries.isEmpty()) {
+
+            return cache.computeIfAbsent(cacheKey, key -> allowAclEntries.stream().anyMatch(matcher));
         } else {
             return false;
         }
     }
 
     public List<AivenAcl> aclEntries() {
-        return List.copyOf(aclEntries);
+        return Stream.of(denyAclEntries, allowAclEntries)
+            .flatMap(List::stream)
+            .collect(Collectors.toList());
     }
```

Comment:
This could be improved a bit. Looking at how this method is used on `Authorizer#acls`, we may just return a Stream instead of a list, as it's turned into a stream _again_ and then mapped to native.

We could even use `Straem.concat` instead:

```
    public Stream<AivenAcl> aclEntries() {
        return Stream.concat(denyAclEntries.stream(), allowAclEntries.stream());
    }
```